esphome/components/template/valve/__init__.py

```diff
@@ -0,0 +1,118 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import automation
+from esphome.components import valve
+from esphome.const import (
+    CONF_ASSUMED_STATE,
+    CONF_CLOSE_ACTION,
+    CONF_CURRENT_OPERATION,
+    CONF_ID,
+    CONF_LAMBDA,
+    CONF_OPEN_ACTION,
+    CONF_OPTIMISTIC,
+    CONF_POSITION,
+    CONF_POSITION_ACTION,
+    CONF_RESTORE_MODE,
+    CONF_STATE,
+    CONF_STOP_ACTION,
+)
+from .. import template_ns
+
+TemplateValve = template_ns.class_("TemplateValve", valve.Valve, cg.Component)
+
+TemplateValveRestoreMode = template_ns.enum("TemplateValveRestoreMode")
+RESTORE_MODES = {
+    "NO_RESTORE": TemplateValveRestoreMode.VALVE_NO_RESTORE,
+    "RESTORE": TemplateValveRestoreMode.VALVE_RESTORE,
+    "RESTORE_AND_CALL": TemplateValveRestoreMode.VALVE_RESTORE_AND_CALL,
+}
+
+CONF_HAS_POSITION = "has_position"
+CONF_TOGGLE_ACTION = "toggle_action"
+
+CONFIG_SCHEMA = valve.VALVE_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(TemplateValve),
+        cv.Optional(CONF_LAMBDA): cv.returning_lambda,
+        cv.Optional(CONF_OPTIMISTIC, default=False): cv.boolean,
+        cv.Optional(CONF_ASSUMED_STATE, default=False): cv.boolean,
+        cv.Optional(CONF_HAS_POSITION, default=False): cv.boolean,
+        cv.Optional(CONF_OPEN_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_CLOSE_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_STOP_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_TOGGLE_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_POSITION_ACTION): automation.validate_automation(single=True),
+        cv.Optional(CONF_RESTORE_MODE, default="RESTORE"): cv.enum(
+            RESTORE_MODES, upper=True
+        ),
```

Comment:
I think a while back we came to a decision that everything should not restore by default and it needs to be user choice.
```suggestion
        cv.Optional(CONF_RESTORE_MODE, default="NO_RESTORE"): cv.enum(
            RESTORE_MODES, upper=True
        ),
```